shared/src/balancer_event_handler.rs

```diff
@@ -0,0 +1,142 @@
+use crate::event_handling::EventIndex;
+use crate::{
+    event_handling::{BlockNumber, EventHandler, EventStoring},
+    impl_event_retrieving,
+    maintenance::Maintaining,
+};
+use anyhow::{anyhow, Context, Result};
+use contracts::{
+    balancer_v2_vault::{
+        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+    },
+    BalancerV2Vault,
+};
+use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160};
+use std::ops::RangeInclusive;
+use tokio::sync::Mutex;
+
+#[derive(Debug)]
+pub enum BalancerEvent {
+    PoolRegistered(PoolRegistered),
+}
+
+#[derive(Debug, Default)]
+pub struct PoolRegistered {
+    pub pool_id: [u8; 32],
+    pub pool_address: H160,
+    pub specialization: u8,
+}
+
+#[derive(Default)]
+pub struct WeightedPool {
+    // pool_address: H160,
+// tokens: Vec<H160>,
+// pool_id:
+// TODO - other stuff
+}
+
+#[derive(Default)]
+pub struct BalancerPools {
+    // pools: HashMap<H160, HashSet<WeightedPool>>,
+// Block number of last update
+// last_updated: u64,
+}
+
+impl BalancerPools {
+    pub fn contract_to_balancer_events(
+        &self,
+        contract_events: Vec<EthContractEvent<ContractEvent>>,
+    ) -> Result<Vec<(EventIndex, BalancerEvent)>> {
+        contract_events
+            .into_iter()
+            .filter_map(|EthContractEvent { data, meta }| {
+                let meta = match meta {
+                    Some(meta) => meta,
+                    None => return Some(Err(anyhow!("event without metadata"))),
+                };
+                match data {
+                    ContractEvent::PoolRegistered(event) => {
+                        Some(convert_pool_registered(&event, &meta))
+                    }
+                    _ => {
+                        tracing::info!("Got {:?}", data);
+                        None
+                    }
+                }
+            })
+            .collect::<Result<Vec<_>>>()
+    }
+}
+
+pub struct BalancerEventUpdater(
+    Mutex<EventHandler<DynWeb3, BalancerV2VaultContract, BalancerPools>>,
+);
+
+impl BalancerEventUpdater {
+    pub fn new(
+        contract: BalancerV2Vault,
+        pools: BalancerPools,
+        start_sync_at_block: Option<u64>,
```

Comment:
This shouldn't be needed as its included in the deployment information of the `BalancerV2Vault`. i.e., you could use `contract.deployment_information()` to get the transaction hash/block number of when the contract got deployed.

Reply:
I am looking at ethcontract docs and I see that this method returns an Option enum:
```
pub enum DeploymentInformation {
    BlockNumber(u64),
    TransactionHash(TransactionHash),
}
```

Meaning that it is optionally block number or transaction hash... I suppose I should handle either case here.

Reply:
Seems that if we wind up with `TransactionHash` at the time this new method is called then we have to make a request to fetch the block number which would make this method async.

Reply:
I have updated this bit here with a somewhat incomplete match statement that leave the `start_sync_at_block` None when the deployment info returns a TransactionHash.

Please take a look and let me know if you had something else in mind.